File: a2a-core-ts/src/json-rpc.ts
import { JsonRpcRequest, JsonRpcResponse, JsonRpcErrorObject } from "./types";

// ── Standard JSON-RPC 2.0 error codes ───────────────────────────────────────

export const JSON_RPC_ERRORS = {
  PARSE_ERROR: { code: -32700, message: "Parse error" },
  INVALID_REQUEST: { code: -32600, message: "Invalid Request" },
  METHOD_NOT_FOUND: { code: -32601, message: "Method not found" },
  INVALID_PARAMS: { code: -32602, message: "Invalid params" },
  INTERNAL_ERROR: { code: -32603, message: "Internal error" },
  // A2A-specific error codes
  TASK_NOT_FOUND: { code: -32001, message: "Task not found" },
  TASK_NOT_CANCELABLE: { code: -32002, message: "Task cannot be canceled" },
  UNSUPPORTED_SKILL: { code: -32003, message: "Unsupported skill" },
} as const;

// ── Request parsing ──────────────────────────────────────────────────────────

export function parseJsonRpcRequest(body: unknown): JsonRpcRequest {
  if (!body || typeof body !== "object") {
    throw makeJsonRpcErrorResponse(null, JSON_RPC_ERRORS.PARSE_ERROR);
  }

  const obj = body as Record<string, unknown>;

  if (obj.jsonrpc !== "2.0") {
    throw makeJsonRpcErrorResponse(
      obj.id as string | number | null,
      JSON_RPC_ERRORS.INVALID_REQUEST,
      'Missing or invalid "jsonrpc" field. Must be "2.0".'
    );
  }

  if (typeof obj.method !== "string" || obj.method.length === 0) {
    throw makeJsonRpcErrorResponse(
      obj.id as string | number | null,
      JSON_RPC_ERRORS.INVALID_REQUEST,
      'Missing or invalid "method" field.'
    );
  }

  if (obj.id === undefined || obj.id === null) {
    throw makeJsonRpcErrorResponse(
      null,
      JSON_RPC_ERRORS.INVALID_REQUEST,
      'Missing "id" field.'
    );
  }

  return {
    jsonrpc: "2.0",
    id: obj.id as string | number,
    method: obj.method as string,
    params: (obj.params as Record<string, unknown>) ?? undefined,
  };
}

// ── Response builders ────────────────────────────────────────────────────────

export function makeJsonRpcResponse(
  id: string | number | null,
  result: unknown
): JsonRpcResponse {
  return {
    jsonrpc: "2.0",
    id: id ?? 0,
    result,
  };
}

export function makeJsonRpcErrorResponse(
  id: string | number | null,
  error: { code: number; message: string },
  data?: unknown
): JsonRpcResponse {
  return {
    jsonrpc: "2.0",
    id: id ?? 0,
    error: {
      code: error.code,
      message: error.message,
      data,
    },
  };
}

File: a2a-core-py/a2a_core/json_rpc.py
"""JSON-RPC 2.0 parsing and response helpers."""

from __future__ import annotations

from typing import Any

from .models import JsonRpcErrorObject, JsonRpcRequest, JsonRpcResponse


# ── Standard error codes ──────────────────────────────────────────────────────

JSON_RPC_ERRORS = {
    "PARSE_ERROR": {"code": -32700, "message": "Parse error"},
    "INVALID_REQUEST": {"code": -32600, "message": "Invalid Request"},
    "METHOD_NOT_FOUND": {"code": -32601, "message": "Method not found"},
    "INVALID_PARAMS": {"code": -32602, "message": "Invalid params"},
    "INTERNAL_ERROR": {"code": -32603, "message": "Internal error"},
    # A2A-specific
    "TASK_NOT_FOUND": {"code": -32001, "message": "Task not found"},
    "TASK_NOT_CANCELABLE": {"code": -32002, "message": "Task cannot be canceled"},
    "UNSUPPORTED_SKILL": {"code": -32003, "message": "Unsupported skill"},
}


# ── Request parsing ──────────────────────────────────────────────────────────

def parse_json_rpc_request(body: dict[str, Any]) -> JsonRpcRequest:
    if not isinstance(body, dict):
        raise ValueError("Request body must be a JSON object")

    if body.get("jsonrpc") != "2.0":
        raise ValueError('Missing or invalid "jsonrpc" field. Must be "2.0".')

    if not isinstance(body.get("method"), str) or len(body["method"]) == 0:
        raise ValueError('Missing or invalid "method" field.')

    if body.get("id") is None:
        raise ValueError('Missing "id" field.')

    return JsonRpcRequest(
        jsonrpc="2.0",
        id=body["id"],
        method=body["method"],
        params=body.get("params"),
    )


# ── Response builders ────────────────────────────────────────────────────────

def make_json_rpc_response(
    rpc_id: str | int | None,
    result: Any,
) -> JsonRpcResponse:
    return JsonRpcResponse(
        jsonrpc="2.0",
        id=rpc_id if rpc_id is not None else 0,
        result=result,
    )


def make_json_rpc_error_response(
    rpc_id: str | int | None,
    error: dict[str, Any],
    data: Any = None,
) -> JsonRpcResponse:
    return JsonRpcResponse(
        jsonrpc="2.0",
        id=rpc_id if rpc_id is not None else 0,
        error=JsonRpcErrorObject(
            code=error["code"],
            message=error["message"],
            data=data,
        ),
    )
